pkg/distributor/distributor_test.go

```diff
@@ -3458,38 +3459,176 @@ func mockWriteRequest(lbls labels.Labels, value float64, timestampMs int64) *mim
+			return 1
+		}
+		return evenlyAssignedZones
+	}
+	return len(c.ingesterStateByZone)
+}
+
+func (c prepConfig) ingesterRingState(zone string, id int) ring.InstanceState {
+	if len(c.ingesterStateByZone[zone].ringStates) == 0 {
+		return ring.ACTIVE
+	}
+	return c.ingesterStateByZone[zone].ringStates[id]
+}
+
+func (c prepConfig) validate(t testing.TB) {
+	if len(c.ingesterStateByZone) != 0 {
+		require.Zero(t, c.numIngesters, "ingesterStateByZone and numIngesters/happyIngesters are exclusive")
+		require.Zero(t, c.happyIngesters, "ingesterStateByZone and numIngesters/happyIngesters are exclusive")
+		require.Nil(t, c.ingesterZones, "ingesterStateByZone and numIngesters/happyIngesters are exclusive")
+
+		for zone, state := range c.ingesterStateByZone {
+			ingestersInZone := state.numIngesters
+			if state.states != nil {
+				require.Zero(t, state.numIngesters, "ingesterStateByZone and numIngesters/happyIngesters are exclusive")
+				require.Zero(t, state.happyIngesters, "ingesterStateByZone and numIngesters/happyIngesters are exclusive")
+				ingestersInZone = len(state.states)
+			}
+			if len(state.ringStates) > 0 {
+				require.Len(t, state.ringStates, ingestersInZone, "ringStates cannot be longer than the number of ingesters in the zone")
+			}
+			require.LessOrEqual(t, len(c.ingesterDataPerZone[zone]), ingestersInZone, "ingesterDataPerZone cannot be longer than the number of ingesters in the zone")
+		}
+	}
+}
+
+func prepareIngesters(cfg prepConfig) []mockIngester {
+	if len(cfg.ingesterStateByZone) != 0 {
+		ingesters := []mockIngester(nil)
+		for zone, state := range cfg.ingesterStateByZone {
+			ingesters = append(ingesters, prepareIngesterZone(zone, state, cfg)...)
+		}
+		return ingesters
+	}
+	ingesters := []mockIngester(nil)
+	numZones := len(cfg.ingesterZones)
+	if numZones == 0 {
+		return prepareIngesterZone("", ingesterZoneState{numIngesters: cfg.numIngesters, happyIngesters: cfg.happyIngesters}, cfg)
+	}
+	for zoneIdx, zone := range cfg.ingesterZones {
+		state := ingesterZoneState{
+			numIngesters:   cfg.numIngesters / numZones,
+			happyIngesters: cfg.happyIngesters / numZones,
+		}
+		if zoneIdx < cfg.happyIngesters%numZones {
+			// Account for cases where the number of happy ingesters isn't divisible by numZones.
+			// For example, when there are 3 zones, 9 ingesters, and 8 happy ingesters.
+			// In this case ingester-zone-c-2 (the last ingester) is the unhappy ingester.
+			state.happyIngesters++
 		}
+
+		if zoneIdx < cfg.numIngesters%numZones {
+			// Account for cases where the number of ingesters isn't divisible by numZones.
+			// For example, when there are 3 zones and 11 ingesters.
+			// In this case the distribution of replicas is zone-a: 4, zone-b: 4, zone-c: 3.
+			state.numIngesters++
+		}
+
+		ingesters = append(ingesters, prepareIngesterZone(zone, state, cfg)...)
+	}
+	return ingesters
+
+}
+
+func prepareIngesterZone(zone string, state ingesterZoneState, cfg prepConfig) []mockIngester {
```

Comment:
[nit] Plural ingesters?

```suggestion
func prepareIngestersZone(zone string, state ingesterZoneState, cfg prepConfig) []mockIngester {
```